### DISCUSSION TITLE
Rebuilding dependencies for Jaylib/raylib?

### DISCUSSION THREAD

**Discussion by oofoe** [janet-lang/janet#1416] [Q&A]
Hi!

I need to do some hacking on raylib (as used by Jaylib) for a little project I'm doing (https://hg.sr.ht/~oofoe/artgrid).

I've pulled Jaylib down into a local tree in my work directory, made the changes in the `%jpm_tree/lib/.cache/git__https___github.com_janet-lang_jaylib/raylib/src/config.h` (I'm just turning on JPG support for now), and then I go to rebuild my main app, but it doesn't recompile raylib.

Is there any way I can get jpm to recompile raylib with my changes? It did it to install it and it would be nice if it could do it again because I just have the MS C++ build tools installed and that doesn't seem to include "make"...

Thanks very much!

**Comment by oofoe** [comment]
So, I went into the cached jaylib source and I can force a rebuild from there (`jpm build`) which does generate new libraries, however they're not installed into the `%jpm_tree/lib/` directory. So I guess I need to figure that out.

I realize that I'm playing with fire a bit, making changes in ephemeral caches, but I'm definitely feeling my way through here without a roadmap.

**Comment by oofoe** [comment]
Well, accidentally did a `jpm install` from the updated cache directory. It installed into my global libraries. However! Now I'm able to use the raylib `%.jpg` support. So I did something right...

Maybe the best way to do it is to clone the jaylib repo, then pull /that/ down and change raylib config.h in it, then install to main janet library directory. I suppose that way I could update my dependencies to use my special jaylib.

**Comment by oofoe** [comment]
OK, made a fork of Jaylib, pulled it down, then had to pull down the raylib stuff that's embedded because I didn't do it recursively. Then modified `raylib/src/config.h` appropriately, did a `jpm build` and `jpm install` and voila! I now have jpeg support when I compile Artgrid. However have to figure out how to put changes back onto github so that other people can build from my modified source, and not sure how recursively included stuff (like raylib) work with a fork?

**Comment by oofoe** [comment]
OK, I have been on a long and arduous journey, but I think I have resolved my issue. At least as good as it's going to get resolved for now...

So the file I need to change is deep inside raylib, which is a submodule of Jaylib. Neither of those projects have extended commit privs to me (probably exercising good judgement, to be honest). This meant that while I could pull them down locally and make the change I needed to `%raylib/src/config.h` and get a library that supported jpeg that I could install, if someone _else_ attempted to build my project from source, it wouldn't have those enhancements.

So, I forked Jaylib and then, because that wasn't _good enough_, I also forked raylib. Then I updated the `%.gitmodules` file in my Jaylib to point to my new raylib and did a `git submodule update --init --recursive`. Then, because submodules are their own repos, I was able to update the magic config file in (my) raylib and bounce back up to (my) Jaylib and do a `jpm build; jpm install`.

Then I went to my Artgrid directory, updated the `%project.janet` file to point to my fork at `https://github.com/oofoe/jaylib` and built _that_. Which has seemed to work (I can load jpeg files). This sure was a lot of work so I could uncomment three lines in an include file though...

BTW, if you want to build Artgrid for your architecture but you already have stock Jaylib installed, then you probably want to be careful that my Jaylib doesn't clobber the other on build (`jpm build -l`?).

**Reply by sogaiu** [reply]
> BTW, if you want to build Artgrid for your architecture but you already have stock Jaylib installed, then you probably want to be careful that my Jaylib doesn't clobber the other on build (jpm build -l?).

Just throwing an idea out here...

I think @saikyun took the path of forking jaylib for his freja project and used the fork as a dependency.  I think evidence of that can be seen in [this line](https://github.com/saikyun/freja/blob/068deb7a3ac1ce405443ebf889b9ed0a643c28e0/project.janet#L26) of the project's `project.janet` file.

The bit that might be slightly different is that I think he changed the name and I think this avoids conflicts with existing `jaylib`s.

**Reply by oofoe** [reply]
Yeah, seems to be the case. However that repo doesn't seem to be available so means that it won't be possible for someone else to rebuild freja completely from source?

I wanted to make it so the Artgrid artifact was completely reproducible by a third party (yes, it's a goofy thing and I can't imagine anyone else would actually _want_ to do it, but it's a good test project for this sort of thing...), so it seemed this was the only way to do it. I'll probably get around to renaming it so it doesn't clobber stuff, but last night I was just tired. ; - )

Thanks for taking a look!

**Reply by sogaiu** [reply]
> I wanted to make it so the Artgrid artifact was completely reproducible by a third party

Oh, I definitely appreciate this.  Thanks for making the effort!

> However that repo doesn't seem to be available so means that it won't be possible for someone else to rebuild freja completely from source?

Actually I believe it is available.  A bit further up in the file is [this line](https://github.com/saikyun/freja/blob/068deb7a3ac1ce405443ebf889b9ed0a643c28e0/project.janet#L6).

**Reply by oofoe** [reply]
OK, that's on me -- I thought I had looked at their repo list, but I was just looking at the "most popular" ones. Thanks for pointing that out! It's interesting to see how someone else modified it.
